actors/eam/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+};
+
+#[cfg(feature = "fil-actor")]
+fil_actors_runtime::wasm_trampoline!(EamActor);
+
+/// Maximum allowed EVM bytecode size.
+/// The contract code size limit is 24kB.
+const MAX_CODE_SIZE: usize = 24 << 10;
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    Create = 2,
+    Create2 = 3,
+    // CreateAccount = 4,
+}
+
+#[derive(Debug)]
+/// Intermediate type between RLP encoding for CREATE
+pub struct RlpCreateAddress {
+    pub address: EthAddress,
+    pub nonce: u64,
+}
+
+impl rlp::Encodable for RlpCreateAddress {
+    fn rlp_append(&self, s: &mut rlp::RlpStream) {
+        s.encoder().encode_value(&self.address.0);
+        s.append(&self.nonce);
+    }
+}
+
+#[derive(serde::Deserialize, serde::Serialize, Debug, Clone, Copy, PartialEq, Eq)]
+pub struct EthAddress(#[serde(with = "strict_bytes")] pub [u8; 20]);
+
+#[derive(Serialize_tuple, Deserialize_tuple)]
+pub struct CreateParams {
+    #[serde(with = "strict_bytes")]
+    pub initcode: Vec<u8>,
+    pub nonce: u64,
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple)]
+pub struct Create2Params {
+    #[serde(with = "strict_bytes")]
+    pub initcode: Vec<u8>,
+    #[serde(with = "strict_bytes")]
+    pub salt: [u8; 32],
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple)]
+pub struct InitAccountParams {
+    #[serde(with = "strict_bytes")]
+    pub pubkey: [u8; SECP_PUB_LEN],
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, PartialEq, Eq)]
+pub struct EamReturn {
+    pub actor_id: ActorID,
+    pub robust_address: Address,
+    pub eth_address: EthAddress,
+}
+
+impl EamReturn {
+    fn from_exec4(exec4: Exec4Return, eth_address: EthAddress) -> Self {
+        Self {
+            actor_id: exec4.id_address.id().unwrap(),
+            robust_address: exec4.robust_address,
+            eth_address,
+        }
+    }
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple, Clone)]
+pub struct EvmConstructorParams {
+    /// The actor's "creator" (specified by the EAM).
+    pub creator: EthAddress,
+    /// The initcode that will construct the new EVM actor.
+    pub initcode: RawBytes,
+}
+
+fn assert_code_size(code: &[u8]) -> Result<(), ActorError> {
+    if code.len() >= MAX_CODE_SIZE {
+        Err(ActorError::illegal_argument(format!(
+            "Supplied EVM initcode {} is larger than max code size 24kB.",
+            code.len()
+        )))
+    } else if code.is_empty() {
+        Err(ActorError::illegal_argument("Supplied EVM initcode cannot be empty.".into()))
+    } else {
+        Ok(())
+    }
+}
+
+/// hash of data with Keccack256, with first 12 bytes cropped
+fn hash_20<BS, RT>(rt: &RT, data: &[u8]) -> [u8; 20]
+where
+    BS: Blockstore + Clone,
+    RT: Runtime<BS>,
+{
+    rt.hash(SupportedHashes::Keccak256, data)[12..32].try_into().unwrap()
+}
+
+fn create_actor<BS, RT>(
+    rt: &mut RT,
+    creator: EthAddress,
+    new_addr: EthAddress,
+    initcode: Vec<u8>,
+) -> Result<RawBytes, ActorError>
```

Comment:
nit: we usually serialize at the very end (right before we return to the system).